markers/env_markers/provider.py

```diff
@@ -152,13 +152,24 @@ def add_prov(prov):
     else:
         allowed_providers = potential_providers
 
+    need_prov_keys = False
+    for filter in filters:
+        if isinstance(filter, ProviderFilter):
+            for filt in filter.classes:
+                if hasattr(filt, 'type_name'):
+                    need_prov_keys = True
+                    break
+
     for provider in allowed_providers:
         argvalues.append([provider])
         # Use the provider key for idlist, helps with readable parametrized test output
         if selector == ONE:
-            idlist.append(provider.category)
-        elif selector == ONE_PER_CATEGORY:
-            idlist.append(provider.type)
+            if need_prov_keys:
+                idlist.append(provider.type_name)
+            else:
+                idlist.append(provider.category)
+        elif selector == ONE_PER_CATEGORY or selector == ONE_PER_TYPE:
```

Comment:
I think this should be divided as in case ONE_PER_CATEGORY will return cloud, infra and as for ONE_PER_TYPE we will get rhos, gce, etc